discretisedfield/plotting/hv.py

```diff
@@ -165,54 +179,48 @@ def __call__(
         vector_kw = {} if vector_kw is None else vector_kw.copy()
 
         if norm_filter or (norm_filter is None and self._norm_filter):
-            if roi is None and "comp" not in self.array.dims:
-                roi = np.abs(self.array)
-            elif roi is None:
-                roi = xr.apply_ufunc(
-                    np.linalg.norm,
-                    self.array,
-                    input_core_dims=[["comp"]],
-                    kwargs={"axis": -1},
-                )
+            if roi is None:
+                roi = self.callback
             scalar_kw.setdefault("roi", roi)
             vector_kw.setdefault("roi", roi)
 
         vector_kw.setdefault("use_color", False)
 
-        if "comp" not in self.array.dims:
+        if "comp" not in self.key_dims:
             return self.scalar(kdims=kdims, **scalar_kw)
-        elif vdims:
-            scalar_dims = list(set(self.array.comp.to_numpy()) - set(vdims))
+
+        # try to guess vdims if not passed
+        if vdims is None and self.vdim_guess_callback is not None:
+            vdims = self.vdim_guess_callback(kdims)
+
+        if vdims:
+            scalar_comps = list(set(self.key_dims["comp"].data) - set(vdims))
             with contextlib.suppress(KeyError):
                 vector_kw.pop("vdims")
             vector = self.vector(kdims=kdims, vdims=vdims, **vector_kw)
-            if len(scalar_dims) == 0:
+            if len(scalar_comps) == 0:
                 return vector
             else:
-                scalar = self.__class__(self.array.sel(comp=scalar_dims)).scalar(
+                key_dims = copy.deepcopy(self.key_dims)
+                if len(scalar_comps) > 1:
+                    key_dims["comp"] = hv_key_dim(scalar_comps, key_dims["comp"].unit)
+                    callback = self.callback
+                else:
+                    # manually remove component 'comp' from returned xarray to avoid
+                    # a drop-down with one element (the out-of-plane component)
+                    def callback(*args, **kwargs):
+                        res = self.callback(*args, **kwargs)
```

Comment:
What is `self.callback()` here? I'm assuming the inner function is not calling itself recursively and i is still the same `self.callback` as was defined in the `__init__()`. Why do you call the inner function also `callback`?

Could you provide an example of `self.callback`?